// @phenomic/plugin-renderer-react/lib/components/Link.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isActive = exports.handleKeyDown = exports.handleClick = exports.handleEvent = undefined;

var _keys = require("babel-runtime/core-js/object/keys");

var _keys2 = _interopRequireDefault(_keys);

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require("babel-runtime/helpers/objectWithoutProperties");

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _typeof2 = require("babel-runtime/helpers/typeof");

var _typeof3 = _interopRequireDefault(_typeof2);

var _react = require("react");

var React = _interopRequireWildcard(_react);

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require("classnames");

var _classnames2 = _interopRequireDefault(_classnames);

var _reactRouter = require("react-router");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var origin = function origin(url) {
  return (typeof url === "undefined" ? "undefined" : (0, _typeof3.default)(url)) === "object" && (
  // jsdom can return "null" string...
  url.origin !== "null" && url.origin ||
  // // IE does not correctly handle origin, maybe Edge does...
  url.protocol + "//" + url.hostname + (url.port ? ":" + url.port : ""));
};

var isSameOrigin = function isSameOrigin(url
// ignore url not from the same domain
// @todo handle sub-folder, see
// https://github.com/phenomic/phenomic/issues/1124
) {
  return origin(url) === origin(window.location);
};

var shouldIgnoreEvent = function shouldIgnoreEvent(event
// If target prop is set (e.g. to "_blank"), let browser handle link.
) {
  return event.currentTarget.target || event.defaultPrevented ||
  // modifier pressed
  event.metaKey || event.altKey || event.ctrlKey || event.shiftKey || false;
};

var goToUrl = function goToUrl(event) {
  if (isSameOrigin(event.currentTarget)) {
    event.preventDefault();
    // extract to get only interesting parts
    var _event$currentTarget = event.currentTarget,
        pathname = _event$currentTarget.pathname,
        search = _event$currentTarget.search,
        hash = _event$currentTarget.hash;

    _reactRouter.browserHistory.push({ pathname: pathname, search: search, hash: hash });
  }
};

var handleEvent = exports.handleEvent = function handleEvent(props, test) {
  return function (event) {
    props && props.onPress && props.onPress(event);
    props && props.onClick && props.onClick(event);
    !shouldIgnoreEvent(event) && (test ? test(event, props) : true) && goToUrl(event);
  };
};

var handleClick = exports.handleClick = function handleClick(props) {
  return handleEvent(props,
  // $FlowFixMe left click
  function (event) {
    return event.button === 0;
  });
};

var handleKeyDown = exports.handleKeyDown = function handleKeyDown(props) {
  return handleEvent(props,
  // $FlowFixMe  enter key
  function (event) {
    return event.keyCode === 13;
  });
};

var isActive = exports.isActive = function isActive(url, _ref) {
  var router = _ref.router;

  var urlObj = { pathname: url };
  // trick to normalize url when possible, by the browser
  // (eg: relative links are "resolved")
  if (typeof document !== "undefined") {
    var link = document.createElement("a");
    link.href = url;
    if (isSameOrigin(link)) {
      urlObj = { pathname: link.pathname };
      // now 'urlObj' is absolute
    }
  }

  return router && router.isActive && router.isActive(urlObj);
};

function Link(props, context) {
  var to = props.to,
      href = props.href,
      style = props.style,
      activeStyle = props.activeStyle,
      className = props.className,
      activeClassName = props.activeClassName,
      otherProps = (0, _objectWithoutProperties3.default)(props, ["to", "href", "style", "activeStyle", "className", "activeClassName"]);

  var url = to || href || "";

  var isUrlActive = isActive(url, context);
  var computedClassName = (0, _classnames2.default)(className, isUrlActive && activeClassName);
  var computedStyle = (0, _extends3.default)({}, style, isUrlActive ? activeStyle : {});
  return React.createElement("a", (0, _extends3.default)({}, otherProps, {
    href: url,
    onClick: handleClick(props),
    onKeyDown: handleKeyDown(props)
    // weird syntax to avoid undefined/empty object/strings
    // for now, it's falling back to normal links
  }, (0, _keys2.default)(computedStyle).length ? { style: computedStyle } : {}, computedClassName ? { className: computedClassName } : {}));
}

Link.contextTypes = {
  router: _propTypes2.default.object.isRequired
};

Link.displayName = "Link";

exports.default = Link;
