// @phenomic/api-client/lib/js/src/phenomicApiClient.js
// Generated by BUCKLESCRIPT VERSION 2.1.0, PLEASE EDIT WITH CARE
'use strict';

var Js_null_undefined = require("bs-platform/lib/js/js_null_undefined.js");

function query(queryConfig) {
  switch (queryConfig.tag | 0) {
    case 0 : 
        var queryConfigItem = queryConfig[0];
        return Object.assign({ }, {
                    path: queryConfigItem[/* path */0],
                    id: queryConfigItem[/* id */1]
                  });
    case 1 : 
        var queryConfigList = queryConfig[0];
        return Object.assign({ }, {
                    path: queryConfigList[/* path */0],
                    by: Js_null_undefined.from_opt(queryConfigList[/* by */1]),
                    value: Js_null_undefined.from_opt(queryConfigList[/* value */2]),
                    order: Js_null_undefined.from_opt(queryConfigList[/* order */3]),
                    limit: Js_null_undefined.from_opt(queryConfigList[/* limit */4])
                  });
    case 2 : 
        var queryConfigPaginatedList = queryConfig[0];
        return Object.assign({ }, {
                    path: queryConfigPaginatedList[/* path */0],
                    by: Js_null_undefined.from_opt(queryConfigPaginatedList[/* by */1]),
                    value: Js_null_undefined.from_opt(queryConfigPaginatedList[/* value */2]),
                    order: Js_null_undefined.from_opt(queryConfigPaginatedList[/* order */3]),
                    limit: Js_null_undefined.from_opt(queryConfigPaginatedList[/* limit */4]),
                    after: Js_null_undefined.from_opt(queryConfigPaginatedList[/* after */5])
                  });
    
  }
}

exports.query = query;
/* No side effect */
