// @phenomic/plugin-transform-json/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _objectWithoutProperties2 = require("babel-runtime/helpers/objectWithoutProperties");

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

exports.default = function () {
  return {
    name: "@phenomic/plugin-transform-json",
    supportedFileTypes: ["json"],
    transform: transformJSONFile
  };
};

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = require("debug")("phenomic:plugin:transform-json");

function transformJSONFile(_ref) {
  var file = _ref.file,
      contents = _ref.contents;

  debug("transforming " + file.name);

  var data = (0, _extends3.default)({}, JSON.parse(contents.toString()), {
    path: _path2.default.basename(file.name, ".json")
  });
  var body = data.body,
      partial = (0, _objectWithoutProperties3.default)(data, ["body"]);


  return {
    data: data,
    partial: partial
  };
}
