// @phenomic/core/lib/api/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decode = exports.encode = undefined;

var _promise = require("babel-runtime/core-js/promise");

var _promise2 = _interopRequireDefault(_promise);

var _slicedToArray2 = require("babel-runtime/helpers/slicedToArray");

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _stringify = require("babel-runtime/core-js/json/stringify");

var _stringify2 = _interopRequireDefault(_stringify);

var _regenerator = require("babel-runtime/regenerator");

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require("babel-runtime/helpers/asyncToGenerator");

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _express = require("express");

var _express2 = _interopRequireDefault(_express);

var _package = require("../../package.json");

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = require("debug")("phenomic:core:api");

var encode = exports.encode = function encode(text) {
  return new Buffer(text).toString("base64");
};
var decode = exports.decode = function decode(text) {
  return new Buffer(text, "base64").toString();
};

var connect = function connect(list, limit) {
  var previousList = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];

  var hasNextPage = limit === undefined ? false : list.length > limit;
  var hasPreviousPage = previousList.length > 0;
  var previousPageIsFirst = limit ? previousList.length <= limit : false;
  // we are retrieving limit + 1 to know if there is more page or not
  // so when getting the previous item, we need to check if we want the last
  // item or the one before (since we added one to the limit)
  var previousIndex = previousList.length - 1 - (previousPageIsFirst ? 0 : 1);
  var nextIndex = list.length - 1;
  return {
    previousPageIsFirst: previousPageIsFirst,
    previous: hasPreviousPage && previousList[previousIndex] ? encode(previousList[previousIndex].id) : undefined,
    next: hasNextPage && list[nextIndex] ? encode(list[nextIndex].id) : undefined,
    list: list.slice(0, limit),
    // deprecated
    hasNextPage: hasNextPage,
    hasPreviousPage: hasPreviousPage
  };
};

function createServer(db, plugins) {
  debug("creating server");
  var server = (0, _express2.default)();

  server.get("/", function () {
    var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(req, res) {
      return _regenerator2.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              debug("get api version");
              res.json({
                engine: "phenomic",
                version: _package2.default.version
              });

            case 2:
            case "end":
              return _context.stop();
          }
        }
      }, _callee, this);
    }));

    return function (_x2, _x3) {
      return _ref.apply(this, arguments);
    };
  }());

  server.get("/:path/by-:filter/:value/:order.json", function () {
    var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(req, res) {
      var reverse, list;
      return _regenerator2.default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              debug(req.url, (0, _stringify2.default)(req.params));
              _context2.prev = 1;
              reverse = req.params.order === "desc";
              _context2.next = 5;
              return db.getList(req.params.path, { reverse: reverse }, req.params.filter, req.params.value);

            case 5:
              list = _context2.sent;

              res.json(connect(list));
              _context2.next = 13;
              break;

            case 9:
              _context2.prev = 9;
              _context2.t0 = _context2["catch"](1);

              console.error(_context2.t0);
              res.status(404).end();

            case 13:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2, this, [[1, 9]]);
    }));

    return function (_x4, _x5) {
      return _ref2.apply(this, arguments);
    };
  }());

  server.get("/:path/by-:filter/:value/:order/limit-:limit.json", function () {
    var _ref3 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(req, res) {
      var limit, reverse, list;
      return _regenerator2.default.wrap(function _callee3$(_context3) {
        while (1) {
          switch (_context3.prev = _context3.next) {
            case 0:
              debug(req.url, (0, _stringify2.default)(req.params));
              _context3.prev = 1;
              limit = parseInt(req.params.limit);
              reverse = req.params.order === "desc";
              _context3.next = 6;
              return db.getList(req.params.path, {
                limit: limit + 1,
                reverse: reverse
              }, req.params.filter, req.params.value);

            case 6:
              list = _context3.sent;

              res.json(connect(list, limit));
              _context3.next = 14;
              break;

            case 10:
              _context3.prev = 10;
              _context3.t0 = _context3["catch"](1);

              console.error(_context3.t0);
              res.status(404).end();

            case 14:
            case "end":
              return _context3.stop();
          }
        }
      }, _callee3, this, [[1, 10]]);
    }));

    return function (_x6, _x7) {
      return _ref3.apply(this, arguments);
    };
  }());

  server.get("/:path/by-:filter/:value/:order/limit-:limit/after-:after.json", function () {
    var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4(req, res) {
      var limit, after, reverse, _ref5, _ref6, list, previousList;

      return _regenerator2.default.wrap(function _callee4$(_context4) {
        while (1) {
          switch (_context4.prev = _context4.next) {
            case 0:
              debug(req.url, (0, _stringify2.default)(req.params));
              _context4.prev = 1;
              limit = parseInt(req.params.limit);
              after = decode(req.params.after);
              // @todo check lt validity (exist?); otherwise, trigger an error (404?)
              // cause during dev all "lt" are responding 200, even random values
              // but in production, it's not the case as only known values are
              // generated as endpoints

              reverse = req.params.order === "desc";
              _context4.next = 7;
              return _promise2.default.all([db.getList(req.params.path, {
                limit: limit + 1,
                gte: after,
                reverse: reverse
              }, req.params.filter, req.params.value), db.getList(req.params.path, {
                limit: limit + 1,
                gt: after,
                reverse: !reverse
              }, req.params.filter, req.params.value)]);

            case 7:
              _ref5 = _context4.sent;
              _ref6 = (0, _slicedToArray3.default)(_ref5, 2);
              list = _ref6[0];
              previousList = _ref6[1];

              res.json(connect(list, limit, previousList));
              _context4.next = 18;
              break;

            case 14:
              _context4.prev = 14;
              _context4.t0 = _context4["catch"](1);

              console.error(_context4.t0);
              res.status(404).end();

            case 18:
            case "end":
              return _context4.stop();
          }
        }
      }, _callee4, this, [[1, 14]]);
    }));

    return function (_x8, _x9) {
      return _ref4.apply(this, arguments);
    };
  }());

  server.get("/:path/item/*.json", function () {
    var _ref7 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee5(req, res) {
      var resource;
      return _regenerator2.default.wrap(function _callee5$(_context5) {
        while (1) {
          switch (_context5.prev = _context5.next) {
            case 0:
              debug(req.url, (0, _stringify2.default)(req.params));
              _context5.prev = 1;
              _context5.next = 4;
              return db.get(req.params.path, req.params["0"]);

            case 4:
              resource = _context5.sent;

              res.json(resource.value);
              _context5.next = 12;
              break;

            case 8:
              _context5.prev = 8;
              _context5.t0 = _context5["catch"](1);

              console.error(_context5.t0);
              res.status(404).end();

            case 12:
            case "end":
              return _context5.stop();
          }
        }
      }, _callee5, this, [[1, 8]]);
    }));

    return function (_x10, _x11) {
      return _ref7.apply(this, arguments);
    };
  }());

  server.get("/item/*.json", function () {
    var _ref8 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee6(req, res) {
      var resource;
      return _regenerator2.default.wrap(function _callee6$(_context6) {
        while (1) {
          switch (_context6.prev = _context6.next) {
            case 0:
              debug(req.url, (0, _stringify2.default)(req.params));
              _context6.prev = 1;
              _context6.next = 4;
              return db.get(null, req.params["0"]);

            case 4:
              resource = _context6.sent;

              res.json(resource.value);
              _context6.next = 12;
              break;

            case 8:
              _context6.prev = 8;
              _context6.t0 = _context6["catch"](1);

              console.error(_context6.t0);
              res.status(404).end();

            case 12:
            case "end":
              return _context6.stop();
          }
        }
      }, _callee6, this, [[1, 8]]);
    }));

    return function (_x12, _x13) {
      return _ref8.apply(this, arguments);
    };
  }());

  // Install the plugins
  plugins.forEach(function (plugin) {
    if (typeof plugin.define === "function") {
      debug("installing plugin '" + plugin.name + "'");
      // $FlowFixMe typeof above is not enough?
      plugin.define(server, db);
    } else {
      debug("plugin '" + plugin.name + "' have no API definition");
    }
  });

  return server;
}

exports.default = createServer;
