// @phenomic/plugin-renderer-react/lib/client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.textRenderer = exports.BodyRenderer = exports.query = exports.Provider = exports.createContainer = exports.createApp = exports.renderApp = undefined;

var _query = require("@phenomic/api-client/lib/query");

var _query2 = _interopRequireDefault(_query);

var _createApp = require("./createApp.js");

var _createApp2 = _interopRequireDefault(_createApp);

var _Container = require("./components/Container");

var _Container2 = _interopRequireDefault(_Container);

var _Provider = require("./components/Provider");

var _Provider2 = _interopRequireDefault(_Provider);

var _BodyRenderer = require("./components/BodyRenderer");

var _BodyRenderer2 = _interopRequireDefault(_BodyRenderer);

var _textRenderer = require("./components/textRenderer");

var _textRenderer2 = _interopRequireDefault(_textRenderer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.renderApp = _createApp.renderApp;
exports.createApp = _createApp2.default;
exports.createContainer = _Container2.default;
exports.Provider = _Provider2.default;
exports.query = _query2.default;
exports.BodyRenderer = _BodyRenderer2.default;
exports.textRenderer = _textRenderer2.default;
