// @phenomic/plugin-renderer-react/lib/renderHTML.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require("babel-runtime/core-js/json/stringify");

var _stringify2 = _interopRequireDefault(_stringify);

var _keys = require("babel-runtime/core-js/object/keys");

var _keys2 = _interopRequireDefault(_keys);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _react = require("react");

var React = _interopRequireWildcard(_react);

var _server = require("react-dom/server");

var _server2 = _interopRequireDefault(_server);

var _HTML = require("./components/HTML");

var _HTML2 = _interopRequireDefault(_HTML);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var debug = require("debug")("phenomic:plugin:react");

/* eslint-disable react/no-multi-comp */
/* eslint-disable react/jsx-no-bind */

var isNotFoundError = function isNotFoundError(e) {
  return e.code === "MODULE_NOT_FOUND";
};

var renderHTML = function renderHTML(_ref) {
  var config = _ref.config,
      props = _ref.props;

  var Html = void 0;
  try {
    // $FlowFixMe Shushhhh!
    Html = require(_path2.default.join(config.path, "Html.js")).default;
  } catch (e) {
    if (!isNotFoundError(e)) {
      throw e;
    }
    debug("Html component cannot be used", e.toString());
    Html = _HTML2.default;
  }

  var base = "/";

  /* eslint-disable react/prop-types */
  return "<!DOCTYPE html>" + _server2.default.renderToStaticMarkup(React.createElement(Html, {
    App: props.WrappedApp,
    render: function render(UserWrappedApp) {
      var _props$renderAsObject = props.renderAsObject(UserWrappedApp),
          main = _props$renderAsObject.main,
          state = _props$renderAsObject.state,
          assets = _props$renderAsObject.assets;

      var sets = (0, _keys2.default)(assets).reduce(function (acc, name) {
        return acc.concat(assets[name]);
      }, []);
      var css = sets.filter(function (asset) {
        return asset.endsWith(".css");
      }).shift();
      var js = sets.filter(function (asset) {
        return asset.endsWith(".js");
      }).shift();
      return {
        html: main,
        Main: function Main(_ref2) {
          var _ref2$html = _ref2.html,
              html = _ref2$html === undefined ? main : _ref2$html;
          return React.createElement("div", {
            id: "PhenomicRoot",
            dangerouslySetInnerHTML: { __html: html || null }
          });
        },
        State: function State() {
          return state && React.createElement("script", {
            id: "PhenomicHydration",
            type: "text/json",
            dangerouslySetInnerHTML: {
              __html: (0, _stringify2.default)(state)
            }
          });
        },
        // eslint-disable-next-line react/no-multi-comp
        Style: function Style() {
          return css ? React.createElement("link", { rel: "stylesheet", href: base + css }) : null;
        },
        Script: function Script() {
          return js ? React.createElement("script", { src: base + js, async: true }) : null;
        },
        assets: assets
      };
    }
  }));
};

exports.default = renderHTML;
