// @phenomic/plugin-renderer-react/lib/shared/store/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require("babel-runtime/helpers/defineProperty");

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _toConsumableArray2 = require("babel-runtime/helpers/toConsumableArray");

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NO_VALUE_EDGE = {
  status: "inactive",
  node: null
};

function createStore() {
  var state = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var subscribers = [];

  function subscribe(func) {
    subscribers = [].concat((0, _toConsumableArray3.default)(subscribers), [func]);
    return function unsubscribe() {
      subscribers = subscribers.filter(function (item) {
        return item !== func;
      });
    };
  }

  function get(key) {
    if (state.hasOwnProperty(key)) {
      return state[key];
    }
    return NO_VALUE_EDGE;
  }

  function set(key, node) {
    update((0, _defineProperty3.default)({}, key, {
      status: "idle",
      node: node
    }));
  }

  function setAsLoading(key) {
    update((0, _defineProperty3.default)({}, key, {
      status: "loading",
      node: get(key).node
    }));
  }

  function setAsError(key, error) {
    update((0, _defineProperty3.default)({}, key, {
      status: "error",
      node: null,
      error: error
    }));
  }

  function update(nextState) {
    state = (0, _extends3.default)({}, state, nextState);
    subscribers.forEach(function (func) {
      return func();
    });
  }

  function getState() {
    return state;
  }

  return {
    subscribe: subscribe,
    get: get,
    set: set,
    setAsLoading: setAsLoading,
    setAsError: setAsError,
    getState: getState
  };
}

exports.default = createStore;
