// @phenomic/plugin-bundler-webpack/lib/webpack.config.js
"use strict";

var _extends2 = require("babel-runtime/helpers/extends");

var _extends3 = _interopRequireDefault(_extends2);

var _defineProperty2 = require("babel-runtime/helpers/defineProperty");

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _webpack = require("webpack");

var _webpack2 = _interopRequireDefault(_webpack);

var _extractTextWebpackPlugin = require("extract-text-webpack-plugin");

var _extractTextWebpackPlugin2 = _interopRequireDefault(_extractTextWebpackPlugin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

module.exports = function (config) {
  return {
    entry: (0, _defineProperty3.default)({}, config.bundleName, [process.env.PHENOMIC_ENV !== "static" && require.resolve("webpack-hot-middleware/client"), process.env.PHENOMIC_ENV !== "static" && require.resolve("react-hot-loader/patch"), "./App.js"].filter(function (item) {
      return item;
    })),
    output: (0, _extends3.default)({
      publicPath: "/", // @todo make this dynamic
      path: _path2.default.isAbsolute(config.outdir) ? config.outdir : _path2.default.join(process.cwd(), config.outdir)
    }, process.env.PHENOMIC_ENV !== "static" ? {
      filename: "phenomic/[name].js",
      chunkFilename: "phenomic/[name].chunk.js"
    } : {
      filename: "phenomic/[name].[chunkhash:8].js",
      chunkFilename: "phenomic/[name].[chunkhash:8].chunk.js"
    }),
    module: {
      rules: [{
        test: /\.js$/,
        exclude: /node_modules/,
        loader: require.resolve("babel-loader"),
        options: {
          babelrc: false,
          presets: [require.resolve("@phenomic/babel-preset")],
          plugins: [require.resolve("react-hot-loader/babel")]
        }
      }, {
        test: /\.css$/,
        loader: _extractTextWebpackPlugin2.default.extract({
          fallback: require.resolve("style-loader"),
          use: require.resolve("css-loader")
        })
      }]
    },
    plugins: [new _extractTextWebpackPlugin2.default({
      filename: "phenomic/[name].[contenthash:8].css",
      disable: process.env.PHENOMIC_ENV !== "static"
    }), process.env.PHENOMIC_ENV !== "static" && new _webpack2.default.HotModuleReplacementPlugin(), process.env.NODE_ENV === "production" && new _webpack2.default.optimize.UglifyJsPlugin()].filter(function (item) {
      return item;
    }),

    resolve: {
      // react-native(-web) | react-primitives
      extensions: [".web.js", ".js", ".json"]
    },

    // eslint-disable-next-line max-len
    // https://github.com/facebookincubator/create-react-app/blob/fbdff9d722d6ce669a090138022c4d3536ae95bb/packages/react-scripts/config/webpack.config.prod.js#L279-L285
    node: {
      fs: "empty",
      net: "empty",
      tls: "empty"
    }
  };
};
