// @phenomic/plugin-renderer-react/lib/server/dev-server.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require("react");

var React = _interopRequireWildcard(_react);

var _server = require("react-dom/server");

var _server2 = _interopRequireDefault(_server);

var _renderHTML = require("../renderHTML");

var _renderHTML2 = _interopRequireDefault(_renderHTML);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var DefaultDevContent = function DefaultDevContent() {
  return (
    /* eslint-disable react-native/no-inline-styles */
    React.createElement(
      "div",
      {
        id: "phenomic-DevLoader",
        style: {
          color: "red",
          font: "caption",
          fontSize: "2rem",
          padding: "40vh 10vw",
          textAlign: "center"
        }
      },
      React.createElement("script", {
        dangerouslySetInnerHTML: {
          __html: "\n    window.onerror = function(e) {\n      var devLoader = document.querySelector(\"#phenomic-DevLoader\")\n      if (devLoader) { devLoader.innerHTML = e.toString() }\n      // only need to use this code once\n      window.onerror = null\n    }"
        }
      }),
      React.createElement(
        "noscript",
        null,
        "Phenomic development server requires JavaScript.\n      If you want to check our your website works without JavaScript, you need to\n      build the static version and server the result.\n      You can do this by doing in your terminal:",
        React.createElement(
          "pre",
          { style: { textAlign: "left" } },
          React.createElement(
            "code",
            null,
            "npm run build" + "\n" + "npm install -g serve" + "\n" + "serve dist"
          )
        )
      )
    )
  );
};
/* eslint-enable react-native/no-inline-styles */

var renderDevServer = function renderDevServer(_ref) {
  var config = _ref.config,
      assets = _ref.assets;
  return (0, _renderHTML2.default)({
    config: config,
    props: {
      WrappedApp: DefaultDevContent,
      renderAsObject: function renderAsObject(UserWrappedApp) {
        return {
          main: process.env.PHENOMIC_ENV === "static" ? _server2.default.renderToString(UserWrappedApp) : _server2.default.renderToStaticMarkup(UserWrappedApp),
          state: null,
          assets: assets
        };
      }
    }
  });
};

exports.default = renderDevServer;
